Split ShorthandSerializer into its own file

darinadler opened the pull request:
#### 4ec904bfc2b063e90dc20b40adfd76ab99f732a2
<pre>
Split ShorthandSerializer into its own file
<a href="https://bugs.webkit.org/show_bug.cgi?id=251064">https://bugs.webkit.org/show_bug.cgi?id=251064</a>
rdar://problem/104586227

Reviewed by Tim Nguyen.

Split the StyleProperties.h/cpp file up so each file has only a single class.
Make a few changes tso there&apos;s enough encapsulation.

* Source/WebCore/Headers.cmake: Added ImmutableStyleProperties.h, MutableStyleProperties.h,
ShorthandSerializer.h, and StylePropertiesInlines.h.

* Source/WebCore/Modules/highlight/Highlight.cpp: Removed unneeded includes.

* Source/WebCore/Sources.txt: Added ImmutableStyleProperties.cpp, MutableStyleProperties.cpp,
and ShorthandSerializer.cpp.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Added all the new files.

* Source/WebCore/animation/DocumentTimeline.cpp: Added an include.

* Source/WebCore/animation/KeyframeEffect.cpp: Added an include.
(WebCore::KeyframeEffect::ParsedKeyframe::ParsedKeyframe): Moved this out of the header so we can
compile without the definition of MutableStyleProperties.
(WebCore::KeyframeEffect::ParsedKeyframe::~ParsedKeyframe): Ditto.

* Source/WebCore/animation/KeyframeEffect.h: Removed include of StyleProperties.h and made
changes so everything can compile without it. Also tidied up a bit.

* Source/WebCore/css/CSSComputedStyleDeclaration.cpp: Updated includes.
* Source/WebCore/css/CSSFontFace.h: Ditto.
* Source/WebCore/css/CSSFontFaceSet.cpp: Ditto.
* Source/WebCore/css/CSSCounterStyleRule.cpp: Ditto.

* Source/WebCore/css/CSSCounterStyleRule.h: Removed include of StyleProperties.h and made
changes so everything can compile without it.
* Source/WebCore/css/CSSKeyframeRule.h: Ditto.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::copyProperties): Renamed from copyPropertiesInSet and updated
to take a Span.
* Source/WebCore/css/ComputedStyleExtractor.h: Ditto.

* Source/WebCore/css/FontSelectionValueInlines.h: Deleted unused isCSS21Weight and
fontWeightKeyword functions.

* Source/WebCore/css/ImmutableStyleProperties.cpp: Added.
* Source/WebCore/css/ImmutableStyleProperties.h: Added.
* Source/WebCore/css/MutableStyleProperties.cpp: Added.
* Source/WebCore/css/MutableStyleProperties.h: Added.
* Source/WebCore/css/ShorthandSerializer.cpp: Added.
* Source/WebCore/css/ShorthandSerializer.h: Added.
* Source/WebCore/css/StyleProperties.cpp: Moved code out of this file.
* Source/WebCore/css/StyleProperties.h: Removed most includes. Marked functions inline that need
inline definitions that depend on the immutable and mutable derived classes. Got rid of
StylePropertiesType and use a boolean m_isMutable instead. Changed copyProperties to take a Span.
* Source/WebCore/css/StylePropertiesInlines.h: Added.

* Source/WebCore/css/StylePropertyShorthand.h:
(WebCore::isShorthandCSSProperty): Deleted. This is now generated along with isLonghand and is
named isShorthand.

* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRule::setProperties): Moved out of line so we can compile the header without
including StyleProperties.h.

* Source/WebCore/css/StyleRule.h: Removed include of StyleProperties.h and made changes so
everything can compile without it.

* Source/WebCore/css/parser/CSSParser.cpp: Updated includes.
* Source/WebCore/css/parser/CSSParserImpl.cpp: Ditto.

* Source/WebCore/css/parser/CSSParserMode.h: Give CSSParserMode an underlying integer type so
it can be forward declared.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValueStart): Call isShorthand.
(WebCore::CSSPropertyParser::parseFontFaceDescriptor): Ditto.

* Source/WebCore/css/process-css-properties.py:
(GenerateCSSPropertyNames): Generate isShorthand alongside isLonghand.

* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp: Updated includes.
(WebCore::CSSStyleValueFactory::parseStyleValue): Use isShorthand.

* Source/WebCore/css/typedom/InlineStylePropertyMap.cpp: Updated includes.

* Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.cpp:
(WebCore::MainThreadStylePropertyMapReadOnly::get const): Use isShorthand.
(WebCore::MainThreadStylePropertyMapReadOnly::getAll const): Ditto.
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::set): Ditto.

* Source/WebCore/dom/Attr.cpp: Updated includes.

* Source/WebCore/dom/ElementData.h: Use MutableStyleProperties instead of StyleProperties for
the types where possible. More specific class gives us more efficient code.

* Source/WebCore/dom/StyledElement.cpp: Updated includes.
* Source/WebCore/dom/StyledElement.h: Use MutableStyleProperties instead of StyleProperties for
the types where possible. More specific class gives us more efficient code.

* Source/WebCore/editing/EditingStyle.cpp: Updated includes.
(WebCore::copyEditingProperties): Use copyProperties.
(WebCore::EditingStyle::extractAndRemoveBlockProperties): Use copyProperties and removeProperties.
(WebCore::EditingStyle::removeBlockProperties): Use removeProperties.
(WebCore::EditingStyle::triStateOfStyle const): Ditto.
(WebCore::removePropertiesInStyle): Ditto.
(WebCore::StyleChange::~StyleChanged): Moved this here sot he header does not need the definition
of StyleProperties.

* Source/WebCore/editing/EditingStyle.h: Removed include of StyleProperties.h and made changes so
everything can compile without it.

* Source/WebCore/editing/Editor.cpp: Updated includes.
* Source/WebCore/editing/EditorCommand.cpp: Ditto.
* Source/WebCore/editing/RemoveFormatCommand.cpp: Ditto.
* Source/WebCore/editing/ReplaceSelectionCommand.cpp: Ditto.
* Source/WebCore/editing/ios/EditorIOS.mm: Ditto.
* Source/WebCore/editing/mac/EditorMac.mm: Ditto.
* Source/WebCore/editing/markup.cpp: Ditto.
* Source/WebCore/html/HTMLBodyElement.cpp: Ditto.
* Source/WebCore/html/HTMLHRElement.cpp: Ditto.
* Source/WebCore/html/HTMLPreElement.cpp: Ditto
.
* Source/WebCore/html/HTMLTableCellElement.cpp:
(WebCore::HTMLTableCellElement::additionalPresentationalHintStyle const): Use
MutableStyleProperties.
* Source/WebCore/html/HTMLTableCellElement.h: Ditto.
* Source/WebCore/html/HTMLTableColElement.cpp:
(WebCore::HTMLTableColElement::additionalPresentationalHintStyle const): Ditto.
* Source/WebCore/html/HTMLTableColElement.h: Ditto.

* Source/WebCore/html/HTMLTableElement.cpp: Updated includes.
(WebCore::leakBorderStyle): Use MutableStyleProperties.
(WebCore::HTMLTableElement::additionalPresentationalHintStyle const): Ditto.
(WebCore::HTMLTableElement::createSharedCellStyle const): Ditto.
(WebCore::HTMLTableElement::additionalCellStyle const): Ditto.
(WebCore::leakGroupBorderStyle): Ditto.
(WebCore::HTMLTableElement::additionalGroupStyle const): Ditto.
* Source/WebCore/html/HTMLTableElement.h: Ditto.
* Source/WebCore/html/HTMLTableSectionElement.cpp:
(WebCore::HTMLTableSectionElement::additionalPresentationalHintStyle const): DItto.
* Source/WebCore/html/HTMLTableSectionElement.h: Ditto.

* Source/WebCore/html/HTMLTextFormControlElement.cpp: Updated includes.
(WebCore::HTMLTextFormControlElement::adjustInnerTextStyle const): Use auto.

* Source/WebCore/html/track/WebVTTParser.cpp: Updated includes.
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp: Ditto.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::buildObjectForAttributesStyle): Use
MutableStyleProperties.

* Source/WebCore/loader/cache/CachedScript.h: Export some symbols used for testing.

* Source/WebCore/page/DragController.cpp: Updated includes.
* Source/WebCore/page/PageSerializer.cpp: Ditto.
* Source/WebCore/platform/animation/TimingFunction.cpp: Ditto.
* Source/WebCore/rendering/RenderTreeAsText.cpp: Ditto.
* Source/WebCore/style/MatchResult.h: Ditto.

* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty): Use isShorthand.
* Source/WebCore/style/Styleable.cpp:
(WebCore::compileTransitionPropertiesInStyle): Ditto.

* Source/WebCore/svg/SVGElementRareData.h: Update includes.
* Source/WebCore/svg/SVGFontFaceElement.cpp: Ditto.
* Source/WebCore/svg/properties/SVGAttributeAnimator.cpp: Ditto.
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageEditorClient.cpp: Ditto.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm: Ditto.
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.mm: Ditto.
* Source/WebKitLegacy/mac/WebView/WebFrame.mm: Ditto.
* Source/WebKitLegacy/mac/WebView/WebView.mm: Ditto.
* Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp: Ditto.

Canonical link: <a href="https://commits.webkit.org/259519@main">https://commits.webkit.org/259519@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/33823b1046122de2b60c3bd136703f75550d7330

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105046 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37936 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114309 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174492 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108949 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15263 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5047 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97365 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113322 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110802 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11802 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94800 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39307 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26427 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80972 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7463 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27786 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7557 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4369 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13612 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47338 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6557 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9346 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on Source/WebCore/css/StyleProperties.h:
```diff
@@ -155,195 +140,26 @@ class StyleProperties : public RefCounted<StyleProperties> {
-    mutable unsigned m_type : 2;
-    unsigned m_arraySize : 27 { 0 };
+    mutable unsigned m_isMutable : 1 { true };
+    unsigned m_arraySize : 28 { 0 };
```
Why does this increase?

darinadler replied:
These bitfields together take 32 bits. We don’t need 2 bits for `type` any more, just 1 bit for `isMutable`. Later maybe we should figure out the real maximum for `arraySize` and set that number of bits precisely. For now, though, as before, it’s just "the rest of the 32 bits".

nt1m commented on Source/WebCore/css/ImmutableStyleProperties.h:
```diff
@@ -0,0 +1,91 @@
+
+    unsigned propertyCount() const { return m_arraySize; }
+    bool isEmpty() const { return !propertyCount(); }
+    PropertyReference propertyAt(unsigned index) const;
```
I haven't checked if there are other members in common between the 2 classes, but is there a reason this needs to be declared here? It's defined in its base class StyleProperties.

Same question for MutableStyleProperties.

darinadler replied:
These are basically virtual functions, although here we implement them manually, to get slightly better performance presumably. The base class checks `isMutable`, and calls the function on either the mutable or immutable derived class. The derived classes have implementations, which can either by called directly or through the base class.

nt1m reviewed (approved):
I wish we had a way for git to record the new files as copies of StyleProperties.cpp for easier review, and to preserve the blame. Though I assume you've left them mostly the same.

darinadler committed:
Split ShorthandSerializer into its own file
https://bugs.webkit.org/show_bug.cgi?id=251064
rdar://problem/104586227

Reviewed by Tim Nguyen.

Split the StyleProperties.h/cpp file up so each file has only a single class.
Make a few changes tso there's enough encapsulation.

* Source/WebCore/Headers.cmake: Added ImmutableStyleProperties.h, MutableStyleProperties.h,
ShorthandSerializer.h, and StylePropertiesInlines.h.

* Source/WebCore/Modules/highlight/Highlight.cpp: Removed unneeded includes.

* Source/WebCore/Sources.txt: Added ImmutableStyleProperties.cpp, MutableStyleProperties.cpp,
and ShorthandSerializer.cpp.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Added all the new files.

* Source/WebCore/animation/DocumentTimeline.cpp: Added an include.

* Source/WebCore/animation/KeyframeEffect.cpp: Added an include.
(WebCore::KeyframeEffect::ParsedKeyframe::ParsedKeyframe): Moved this out of the header so we can
compile without the definition of MutableStyleProperties.
(WebCore::KeyframeEffect::ParsedKeyframe::~ParsedKeyframe): Ditto.

* Source/WebCore/animation/KeyframeEffect.h: Removed include of StyleProperties.h and made
changes so everything can compile without it. Also tidied up a bit.

* Source/WebCore/css/CSSComputedStyleDeclaration.cpp: Updated includes.
* Source/WebCore/css/CSSFontFace.h: Ditto.
* Source/WebCore/css/CSSFontFaceSet.cpp: Ditto.
* Source/WebCore/css/CSSCounterStyleRule.cpp: Ditto.

* Source/WebCore/css/CSSCounterStyleRule.h: Removed include of StyleProperties.h and made
changes so everything can compile without it.
* Source/WebCore/css/CSSKeyframeRule.h: Ditto.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::copyProperties): Renamed from copyPropertiesInSet and updated
to take a Span.
* Source/WebCore/css/ComputedStyleExtractor.h: Ditto.

* Source/WebCore/css/FontSelectionValueInlines.h: Deleted unused isCSS21Weight and
fontWeightKeyword functions.

* Source/WebCore/css/ImmutableStyleProperties.cpp: Added.
* Source/WebCore/css/ImmutableStyleProperties.h: Added.
* Source/WebCore/css/MutableStyleProperties.cpp: Added.
* Source/WebCore/css/MutableStyleProperties.h: Added.
* Source/WebCore/css/ShorthandSerializer.cpp: Added.
* Source/WebCore/css/ShorthandSerializer.h: Added.
* Source/WebCore/css/StyleProperties.cpp: Moved code out of this file.
* Source/WebCore/css/StyleProperties.h: Removed most includes. Marked functions inline that need
inline definitions that depend on the immutable and mutable derived classes. Got rid of
StylePropertiesType and use a boolean m_isMutable instead. Changed copyProperties to take a Span.
* Source/WebCore/css/StylePropertiesInlines.h: Added.

* Source/WebCore/css/StylePropertyShorthand.h:
(WebCore::isShorthandCSSProperty): Deleted. This is now generated along with isLonghand and is
named isShorthand.

* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRule::setProperties): Moved out of line so we can compile the header without
including StyleProperties.h.

* Source/WebCore/css/StyleRule.h: Removed include of StyleProperties.h and made changes so
everything can compile without it.

* Source/WebCore/css/parser/CSSParser.cpp: Updated includes.
* Source/WebCore/css/parser/CSSParserImpl.cpp: Ditto.

* Source/WebCore/css/parser/CSSParserMode.h: Give CSSParserMode an underlying integer type so
it can be forward declared.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValueStart): Call isShorthand.
(WebCore::CSSPropertyParser::parseFontFaceDescriptor): Ditto.

* Source/WebCore/css/process-css-properties.py:
(GenerateCSSPropertyNames): Generate isShorthand alongside isLonghand.

* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp: Updated includes.
(WebCore::CSSStyleValueFactory::parseStyleValue): Use isShorthand.

* Source/WebCore/css/typedom/InlineStylePropertyMap.cpp: Updated includes.

* Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.cpp:
(WebCore::MainThreadStylePropertyMapReadOnly::get const): Use isShorthand.
(WebCore::MainThreadStylePropertyMapReadOnly::getAll const): Ditto.
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::set): Ditto.

* Source/WebCore/dom/Attr.cpp: Updated includes.

* Source/WebCore/dom/ElementData.h: Use MutableStyleProperties instead of StyleProperties for
the types where possible. More specific class gives us more efficient code.

* Source/WebCore/dom/StyledElement.cpp: Updated includes.
* Source/WebCore/dom/StyledElement.h: Use MutableStyleProperties instead of StyleProperties for
the types where possible. More specific class gives us more efficient code.

* Source/WebCore/editing/EditingStyle.cpp: Updated includes.
(WebCore::copyEditingProperties): Use copyProperties.
(WebCore::EditingStyle::extractAndRemoveBlockProperties): Use copyProperties and removeProperties.
(WebCore::EditingStyle::removeBlockProperties): Use removeProperties.
(WebCore::EditingStyle::triStateOfStyle const): Ditto.
(WebCore::removePropertiesInStyle): Ditto.
(WebCore::StyleChange::~StyleChanged): Moved this here sot he header does not need the definition
of StyleProperties.

* Source/WebCore/editing/EditingStyle.h: Removed include of StyleProperties.h and made changes so
everything can compile without it.

* Source/WebCore/editing/Editor.cpp: Updated includes.
* Source/WebCore/editing/EditorCommand.cpp: Ditto.
* Source/WebCore/editing/RemoveFormatCommand.cpp: Ditto.
* Source/WebCore/editing/ReplaceSelectionCommand.cpp: Ditto.
* Source/WebCore/editing/ios/EditorIOS.mm: Ditto.
* Source/WebCore/editing/mac/EditorMac.mm: Ditto.
* Source/WebCore/editing/markup.cpp: Ditto.
* Source/WebCore/html/HTMLBodyElement.cpp: Ditto.
* Source/WebCore/html/HTMLHRElement.cpp: Ditto.
* Source/WebCore/html/HTMLPreElement.cpp: Ditto
.
* Source/WebCore/html/HTMLTableCellElement.cpp:
(WebCore::HTMLTableCellElement::additionalPresentationalHintStyle const): Use
MutableStyleProperties.
* Source/WebCore/html/HTMLTableCellElement.h: Ditto.
* Source/WebCore/html/HTMLTableColElement.cpp:
(WebCore::HTMLTableColElement::additionalPresentationalHintStyle const): Ditto.
* Source/WebCore/html/HTMLTableColElement.h: Ditto.

* Source/WebCore/html/HTMLTableElement.cpp: Updated includes.
(WebCore::leakBorderStyle): Use MutableStyleProperties.
(WebCore::HTMLTableElement::additionalPresentationalHintStyle const): Ditto.
(WebCore::HTMLTableElement::createSharedCellStyle const): Ditto.
(WebCore::HTMLTableElement::additionalCellStyle const): Ditto.
(WebCore::leakGroupBorderStyle): Ditto.
(WebCore::HTMLTableElement::additionalGroupStyle const): Ditto.
* Source/WebCore/html/HTMLTableElement.h: Ditto.
* Source/WebCore/html/HTMLTableSectionElement.cpp:
(WebCore::HTMLTableSectionElement::additionalPresentationalHintStyle const): DItto.
* Source/WebCore/html/HTMLTableSectionElement.h: Ditto.

* Source/WebCore/html/HTMLTextFormControlElement.cpp: Updated includes.
(WebCore::HTMLTextFormControlElement::adjustInnerTextStyle const): Use auto.

* Source/WebCore/html/track/WebVTTParser.cpp: Updated includes.
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp: Ditto.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::buildObjectForAttributesStyle): Use
MutableStyleProperties.

* Source/WebCore/loader/cache/CachedScript.h: Export some symbols used for testing.

* Source/WebCore/page/DragController.cpp: Updated includes.
* Source/WebCore/page/PageSerializer.cpp: Ditto.
* Source/WebCore/platform/animation/TimingFunction.cpp: Ditto.
* Source/WebCore/rendering/RenderTreeAsText.cpp: Ditto.
* Source/WebCore/style/MatchResult.h: Ditto.

* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty): Use isShorthand.
* Source/WebCore/style/Styleable.cpp:
(WebCore::compileTransitionPropertiesInStyle): Ditto.

* Source/WebCore/svg/SVGElementRareData.h: Update includes.
* Source/WebCore/svg/SVGFontFaceElement.cpp: Ditto.
* Source/WebCore/svg/properties/SVGAttributeAnimator.cpp: Ditto.
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageEditorClient.cpp: Ditto.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm: Ditto.
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.mm: Ditto.
* Source/WebKitLegacy/mac/WebView/WebFrame.mm: Ditto.
* Source/WebKitLegacy/mac/WebView/WebView.mm: Ditto.
* Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp: Ditto.

Canonical link: https://commits.webkit.org/259519@main

weinig commented on Source/WebCore/css/ImmutableStyleProperties.cpp:
```diff
@@ -0,0 +1,88 @@
+ImmutableStyleProperties::~ImmutableStyleProperties()
+{
+    auto* valueArray = bitwise_cast<PackedPtr<CSSValue>*>(this->valueArray());
+    for (unsigned i = 0; i < m_arraySize; ++i)
```
Is there a particular reason this is using a traditional index based rather than range-for?

weinig replied:
(It might be as simple as that this is code that you are moving and I didn't upgrade this to a range loop when I added support for iterators here).

darinadler replied:
The function `valueArray` returns a pointer. We could change that to a span.

weinig commented on Source/WebCore/css/ImmutableStyleProperties.cpp:
```diff
@@ -0,0 +1,88 @@
+{
+    auto* metadataArray = const_cast<StylePropertyMetadata*>(this->metadataArray());
+    auto* valueArray = bitwise_cast<PackedPtr<CSSValue>*>(this->valueArray());
+    for (unsigned i = 0; i < length; ++i) {
```
Man, a ZipperedIterator would be nice here:

```c++
    auto* metadataArray = const_cast<StylePropertyMetadata*>(this->metadataArray());
    auto* valueArray = bitwise_cast<PackedPtr<CSSValue>*>(this->valueArray());
    for (auto& [metadata, value, property] : WTF::zippered(metadataArray, valueArray, properties)) {
         metadata = property.metadata();
         value = property.value();
         value->ref();
    }
```

Maybe it's time to add one. Seems like a fun weekend project for me - https://bugs.webkit.org/show_bug.cgi?id=251319.

darinadler replied:
Yes, this would work if we switch from pointers to spans. Might have to rejigger the `const` handling too.

weinig commented on Source/WebCore/css/ImmutableStyleProperties.cpp:
```diff
@@ -0,0 +1,88 @@
+    // Convert here propertyID into an uint16_t to compare it with the metadata's m_propertyID to avoid
+    // the compiler converting it to an int multiple times in the loop.
+    uint16_t id = static_cast<uint16_t>(propertyID);
+    for (int n = m_arraySize - 1 ; n >= 0; --n) {
```
I can't recall what the purpose of iterating backwards is here. Do you remember? Maybe it would be good to add a comment at some point.

Loirooriol replied:
My guess would be that, if by mistake the property appears twice, we want the later one, since it takes precedence.

darinadler replied:
That guess sounds right to me. CSS has "last value wins". Might be better to put that in the function’s name. Call it `findLast` instead of just `find`.

weinig commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -0,0 +1,1175 @@
+    return serializeLonghands(length() - isLonghandInitialValue(length() - 1), separator);
+}
+
+inline String ShorthandSerializer::serializeCommonValue() const
```
Does this `inline` do anything?

darinadler replied:
It does not. I gave in to outdated urges and put in quite a few `inline` like this one that likely have no effect, but my meaning was "I hope this gets inlined".
